nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+    std::bind(&CollisionMonitorNode::publishPolygons, this));
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Deactivating main worker
+  setWorkerActive(false);
+
+  // Reset opeartion states to their defaults after worker deactivating
+  release_operation_ = false;
+  ra_prev_ = {DO_NOTHING, {0.0, 0.0, 0.0}};
+
+  // Deactivating lifecycle publishers
+  polygon_pub_->on_deactivate();
+  cmd_vel_out_pub_->on_deactivate();
+
+  // Deactivating polygons publishing timer
+  polygon_pub_timer_->cancel();
+
+  // Destroying bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  odom_sub_.reset();
+  cmd_vel_in_sub_.reset();
+
+  polygon_pub_.reset();
+  cmd_vel_out_pub_.reset();
+
+  polygon_pub_timer_.reset();
+
+  polygons_.clear();
+  sources_.clear();
+
+  resetVelocity();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void CollisionMonitorNode::publishPolygons()
+{
+  if (!polygon_topic_.length()) {
+    // Topic is not specified: there is no need to publish
+    return;
+  }
+
+  // Fill PolygonStamped struct
+  std::unique_ptr<geometry_msgs::msg::PolygonStamped> poly_s =
+    std::make_unique<geometry_msgs::msg::PolygonStamped>();
+  poly_s->header.stamp = this->now();
+  poly_s->header.frame_id = base_frame_id_;
+  poly_s->polygon = polygons_pub_;
+
+  // Publish polygon
+  polygon_pub_->publish(std::move(poly_s));
+}
+
+void CollisionMonitorNode::odomCallback(nav_msgs::msg::Odometry::ConstSharedPtr msg)
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
+
+  double dt;
+  try {
+    dt = (this->now() - msg->header.stamp).seconds();
+  } catch (const std::exception & ex) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Can not get difference between collision monitor and odom time: %s",
+      ex.what());
+    dt = 0.0;
+  }
+
+  if (std::fabs(dt) <= max_time_shift_) {
+    velocity_.x = msg->twist.twist.linear.x;
+    velocity_.y = msg->twist.twist.linear.y;
+    velocity_.tw = msg->twist.twist.angular.z;
+    velocity_valid_ = true;
+  } else {
+    // Warn and ignore the odom data if it and Collision Monitor node have different time stamps
+    RCLCPP_WARN(
+      get_logger(),
+      "Odom and collision monitor node timestamps differ on %f seconds. Ignoring odom.",
+      dt);
+    velocity_valid_ = false;
+  }
+}
+
+Velocity CollisionMonitorNode::getVelocity()
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
```

Comment:
I largely don't think most of the mutexes are required. All of the threads are being processed on the single threaded executor queue, so they can't run at the same time unless we are manually setting a `std::thread` separate of the timer/subscriber callbacks the rclcpp queue is handling for us. It is promised to never be running at the same time.

Reply:
ROS 2 has its own callback queue in the executor, so if a node is given a single threaded executor, it can only process a single thing at a time, guaranteeing no async issues. It only becomes a problem with multiple executors or multithreaded executors (or things happening outside of ROS' control)